### ISSUE TITLE
ServeDir: requests that ask for a range that extends past the end of a document return a 416 response

### ISSUE THREAD

**Issue by jfaust** [tower-rs/tower-http#384]
### Version

`0.4.1`

### Description

This is... not necessarily a bug report so much as a compatibility report? I was just working on getting some SEO things working on my site, only to discover that LinkedIn's post inspector fails (https://www.linkedin.com/post-inspector/). This is because their scraper send a `range: bytes=0-3145727` header no matter what. I'm assuming this is so they only pull up to 3MB of the page... however, it causes `ServeDir` to return a 416 (Range Not Satisfiable). On the face of it this seems like the correct behavior for `ServeDir`. However, I've just tested a few http servers (using `curl`), and so far they're all happy just returning the full document in this case.

Thoughts?

**Comment by jfaust** [comment]
FYI, not an urgent question, as I've written some middleware to strip the range header in this case. Just thought it'd be worth bringing up here.

**Comment by davidpdrsn** [comment]
Yes I think so

**Comment by jfaust** [comment]
This turns out a little more complicated than I was expecting, because the actual error is getting raised by `http-range-header::ParsedRanges::validate()`. 

Does this change belong there? I didn't originally think so, but now I'm not sure. Reading https://www.ietf.org/rfc/rfc2616.txt, section `10.4.17`:
```
10.4.17 416 Requested Range Not Satisfiable

   A server SHOULD return a response with this status code if a request
   included a Range request-header field (section 14.35), and none of
   the range-specifier values in this field overlap the current extent
   of the selected resource, and the request did not include an If-Range
   request-header field. (For byte-ranges, this means that the first-
   byte-pos of all of the byte-range-spec values were greater than the
   current length of the selected resource.)

   When this status code is returned for a byte-range request, the
   response SHOULD include a Content-Range entity-header field
   specifying the current length of the selected resource (see section
   14.16). This response MUST NOT use the multipart/byteranges content-
   type.
```

`none of the range-specifier values in this field overlap` makes me think that the behavior of `http-range-header` is wrong in this case, but I could be parsing it wrong. I also haven't read the entire RFC, so I may be missing other context.

If `http-range-header` is right, and we just want to be lax in `ServeDir`, then I think I need to basically duplicate `ParsedRanges::validate()` into `tower-http`. I'll file a ticket with https://github.com/MarcusGrass/http-range-header and see what he thinks.

**Comment by seanmonstar** [comment]
Here's the latest version of the relevant part, [RFC 9110 Section 14.1.2](https://httpwg.org/specs/rfc9110.html#rfc.section.14.1.2):

> If the [last-pos](https://httpwg.org/specs/rfc9110.html#rule.int-range) value is absent, or if the value is greater than or equal to the current length of the representation data, the byte range is interpreted as the remainder of the representation (i.e., the server replaces the value of [last-pos](https://httpwg.org/specs/rfc9110.html#rule.int-range) with a value that is one less than the current length of the selected representation).

**Comment by jfaust** [comment]
@seanmonstar Ah, perfect. Thanks

**Comment by jfaust** [comment]
I've opened a PR with `http-range-header`: https://github.com/MarcusGrass/http-range-header/pull/3

**Comment by MarcusGrass** [comment]
Fixed up and published under `0.3.1`, on second thoughts that might have been dubious but I'm filing this under a bug-fix/patch!
